myst_nbs/time_series/Forecasting_with_structural_timeseries.myst.md

```diff
@@ -0,0 +1,662 @@
+
+fig, ax = plt.subplots(figsize=(10, 3))
+ax.set_title("Generated Autoregressive Timeseries", fontsize=15)
+ax.plot(ar1_data);
+```
+
+## Specifying the Model
+
+We'll walk through the model step by step and then generalise the pattern into a function that can be used to take increasingly complex structural combinations of components.
+
+```{code-cell} ipython3
+## Set up a dictionary for the specification of our priors
+## We set up the dictionary to specify size of the AR coefficients in
+## case we want to vary the AR lags.
+priors = {"coefs": {"mu": 0, "size": 2}, "sigma": 1, "init": {"mu": 8, "sigma": 1, "size": 1}}
+
+## Initialise the model
+with pm.Model() as AR:
+    pass
+
+## Define the time interval for fitting the data
+t_data = list(range(len(ar1_data)))
+## Add the time interval as a mutable coordinate to the model to allow for future predictions
+AR.add_coord("obs_id", t_data, mutable=True)
+
+with AR:
+    ## Data containers to enable prediction
+    t = pm.MutableData("t", t_data, dims="obs_id")
+    y = pm.MutableData("y", ar1_data, dims="obs_id")
+
+    # The first coefficient will be the constant term but we need to set priors for each coefficient in the AR process
+    coefs = pm.Normal("coefs", priors["coefs"]["mu"], size=priors["coefs"]["size"])
+    sigma = pm.HalfNormal("sigma", priors["sigma"])
+    # We need one init variable for each lag, hence size is variable too
+    init = pm.Normal.dist(
+        priors["init"]["mu"], priors["init"]["sigma"], size=priors["init"]["size"]
+    )
+    # Steps of the AR model minus the lags required
+    ar1 = pm.AR(
+        "ar",
+        coefs,
+        sigma=sigma,
+        init_dist=init,
+        constant=True,
+        steps=t.shape[0] - (priors["coefs"]["size"] - 1),
+    )
+
+    # The Likelihood
+    outcome = pm.Normal("likelihood", mu=ar1, sigma=sigma, observed=y)
+    ## Sampling
+    idata_ar = pm.sample_prior_predictive()
+    idata_ar.extend(pm.sample(2000, random_seed=100, target_accept=0.95))
+    idata_ar.extend(pm.sample_posterior_predictive(idata_ar))
+```
+
+```{code-cell} ipython3
+idata_ar
+```
+
+Lets check the model structure with plate notation and then examine the convergence diagnostics.
+
+```{code-cell} ipython3
+pm.model_to_graphviz(AR)
+```
+
+```{code-cell} ipython3
+az.plot_trace(idata_ar, figsize=(10, 6), kind="rank_vlines");
+```
+
+Next we'll check the summary estimates for the to AR coefficients and the sigma term.
+
+```{code-cell} ipython3
+az.summary(idata_ar, var_names=["~ar"])
+```
+
+## Prediction Step
+
+The next step works much like generating posterior predictive observations for new data in a GLM model but with one additional complication that the AR process also expects "new" data. Or put another way, we have to tell the model how many prediction steps we want to impute with the model we have just fit. So for the shape handling purposes we have to feed our model the observed data and an appropriate range of null values.
```

Comment:
This step is super important but I’m sorry to have to tell you that there is something wrong. This had come up before in a thread in [discourse](https://discourse.pymc.io/t/making-out-of-sample-predictions-with-gaussianrandomwalk/9703/4?u=lucianopaz), but we never found the time to move that into the pymc documentation. I think that your PR is the perfect opportunity to finally add a good explanation on how to do forecasts of time series.
To summarise what was said in the thread, forecasting a time series is not the same as doing predictions with a GLM. The reason these are different is that a time series is conditionally dependent on the past values of the series, whilst the values of a GLM are conditionally independent from each other (given the parameters of the GLM).
The correct way to do a forecast of a time series is to create a new random variable using they learnt AR coefficients, but setting the `init` to a `DiracDelta` of the last entries of the past series. If this isn’t clear, I’m happy to clarify it and provide examples on how this could be done.
@ricardoV94  has a very nice gist that concatenates two random variables, one defined on constant data, and the other on mutable data, which could be a very nice way to do this here.
The other subtlety of forecasting a time series is whether the series is latent or if it was observed. The only real distinction between the two cases is whether the forecast’s init should be defined in terms of the AR random variable, or in terms of its observations directly.

Reply:
Yes, so i've re-run the notebook with a different pattern of predict step based on the above. Not sure it's exactly what you had in mind, but it seems to work. Base pattern is like this: 

<img width="951" alt="image" src="https://user-images.githubusercontent.com/10588167/197362951-f140d34d-84e2-47c3-9d60-dc0e94cbfe50.png">

The model structure changes here too, so as to emphasise the conditional nature of the forecast. I think this is big improvement on the prior method and really appreciate the pointer, but do let me know if you had something else in mind?

<img width="676" alt="image" src="https://user-images.githubusercontent.com/10588167/197362999-2e2e3a54-42d6-4f9e-8c31-405c2c9c3c90.png">

With similar adjustments made for the trend and seasonal constructs. 

Plotting still roughly works the same way: 

<img width="1117" alt="image" src="https://user-images.githubusercontent.com/10588167/197363034-e75e8efc-9fa6-451c-be75-c5ab40908946.png">